include/deal.II/base/lazy.h

```diff
@@ -0,0 +1,215 @@
+#define dealii_initialization_check_h
+
+
+#include <deal.II/base/config.h>
+
+#include <mutex>
+#include <optional>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * @addtogroup threads
+ * @{
+ */
+
+/**
+ * This class is a thin wrapper around std::optional that provides a
+ * convenient mechanism for lazy initialization of the contained object on
+ * first use. The class ensures that on-demand initialization of some
+ * expensive data structure happens (a) in a thread-safe manner, and that
+ * (b) subsequent checks in hot paths are cheap.
+ *
+ * Example usage:
+ * ```
+ * template<...>
+ * class FE<...>
+ * {
+ * public:
+ *   const FullMatrix<double> &
+ *   get_prolongation_matrix() const
+ *   {
+ *     prolongation_matrix.initialize([&](){
+ *       // ... Some expensive operation initializing the prolongation
+ *       // matrix that we only want to perform once and when necessary.
+ *       });
+ *     return prolongation_matrix;
+ *   }
+ *
+ * private:
+ *   mutable Lazy<FullMatrix<double>> prolongation_matrix;
+ * }
+ * ```
+ *
+ * @note This class inherits from std::optional<T> which implies that most
+ * of the public interface of std::optional is available.
+ */
+template <typename T>
+class Lazy : public std::optional<T>
+{
+public:
+  /**
+   * Default Constructor.
+   */
+  Lazy() = default;
+
+
+  /**
+   * Copy constructor.
+   */
+  Lazy(const Lazy &other)
+    : std::optional<T>(other)
+  {
+    //
+    // The only legal operation for a std::once_flag is to default
+    // construct it which sets its internal state to "function not
+    // called". This is OK as we first check the std::optional<T> whether
+    // it contains a value.
+    //
+  }
+
+
+  /**
+   * Move constructor.
+   */
+  Lazy(Lazy &&other) noexcept
+    : std::optional<T>(other)
+  {
+    //
+    // The only legal operation for a std::once_flag is to default
+    // construct it which sets its internal state to "function not called".
+    // This is OK as we first check the std::optional<T> whether it
+    // contains a value.
+    //
+  }
+
+
+  /**
+   * Copy assignment.
+   */
+  Lazy &
+  operator=(const Lazy &other)
+  {
+    static_cast<std::optional<T> &>(*this) =
+      static_cast<const std::optional<T> &>(other);
+
+    //
+    // Copy assignment might reset the std::optional<T> with an empty one
+    // containing no value. Thus, we have to reset the std::once_flag to
+    // "function not called".
+    //
+    flag.~once_flag();
+    new (&flag) std::once_flag;
+
+    return *this;
+  }
+
+
+  /**
+   * Move assignment.
+   */
+  Lazy &
+  operator=(Lazy &&other) noexcept
+  {
+    static_cast<std::optional<T> &>(*this) =
+      static_cast<std::optional<T> &&>(other);
+
+    //
+    // Copy assignment might reset the std::optional<T> with an empty one
+    // containing no value. Thus, we have to reset the std::once_flag to
+    // "function not called".
+    //
+    flag.~once_flag();
+    new (&flag) std::once_flag;
+
+    return *this;
```

Comment:
I think that's wrong. The right code seems to me like this:
```
  if (this->has_value())
    // value has already been computed. We need not run the initialization again, so invalidate the flag if it isn't already
    {
      std::call_once(flag, [](){});
    }
  else
    {
      flag.~once_flag();
      new (&flag) std::once_flag;
    }
```

Reply:
This is a subtle point - I will try to document it a bit better: If the optional contains a value (after copy/move assignment) we don't ever need to set `flag` to the initialized state again: In this case `has_value()` returns true and we simply never reach the branch with the `std::call_once` call. The flag is only ever needed to synchronize the `std::call_once` for the small window when the `std::optional` payload gets initialized.